### PR TITLE
Multiemail field issues

### PR DESCRIPTION
**Pull request by davide-alghi**
**[Bug] Multiemail field issues**

Before explain issues, I have a question: what's better?

1) I create directly a pull request with issue explanation
2) Before pull request creation, I have to create an issue and then I create a pull request
3) I create an issue and wait your agree about pull request creation

=================================== ** ===================================

I tested MultiEmail field type and I found some issues.
multiemailissues
multiemailnewticked

### THE BUGS =====================================

**=== Editview**
1) 'Consent to send' label overlaps partially the checkbox (and fa-square) - because label text wraps
2) plus-button adds a new email box with ticked consent checkbox if there is already a ticked consent checkbox - because in MultiEmail.addEmail (/public_html/layouts/resources/Fields.js) virtual checkbox tick is not removed before adding
3) on 'Consent to send' click, checkbox is not ticked and viceversa - because in MultiEmail.toggleCheckBox (/public_html/layouts/resources/Fields.js) 'svg.svg-inline--fa' doesn'e exist, so it fails
4) on window resize, field overflows and 'Consent to send' checkbox could not be accessible - because in /public_html/layouts/basic/styles/Main.css 'form-group' class has "flex: 0 0 auto"

**=== DetailView**
1) consent is not available: no info about consent for single email
2) I can send email by click if consent is negated
3) inline editing doesn't work: add and remove buttons don't work, saving don't work (outer click) - add and remove don't work because in /public_html/layouts/basic/modules/Vtiger/resources/Detail.js:registerBasicEvents() doesn't call App.Fields.MultiEmail.register(detailContentsHolder), so events handlers are not binded. "save" doesn't work because validation fails always in saveHandler function: fieldElement is null; there should be needed a loop to validate every single email field.

**=== XML EXPORT**
1) field is json: in xml is html encoded

### MY OPINIONS =====================================

**=== Editview**
1) layout should have to be restyled: label without space wrapping and checkbox on the square bottom.
4) standard flex is "0 1 auto": I think this is better for browser compatibility and RWD

**=== DetailView**
1. => 2.
2. this information should have to be available in detail: email, with consent, clickable and with a little green tick; email without consent not-clickable and with a close red lock (in listview and excelExport only text, no tags)
3. I think that the quick fix is to disable AjaxEditable for the field

**=== XML EXPORT**
1) for uniformity reason (respect other json values), value should be in CDATA format - MultiEmail field type is the only one with json content (inventory excluded)

### FIXES =====================================

**=== Editview**

1) in /public_html/layouts/basic/styles/Main.css (.c-float-label__container and c-float-label__label classes are used only in MultiEmail) set -ms-flex and flex to "0 1 auto"
```CSS
.form-inline .form-group {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 0;
    -ms-flex: 0 1 auto;
        flex: 0 1 auto;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
    -ms-flex-flow: row wrap;
      flex-flow: row wrap;
  -webkit-box-align: center;
    -ms-flex-align: center;
      align-items: center;
  margin-bottom: 0;
}
```
added to move all on bottom
```CSS
.c-float-label__container {
  top: 0.6em;}
```
and removed "left: 0" to center label in overflow case; added "white-space: nowrap" to keep label on single line
```CSS
.c-float-label__label {
  position: absolute;
  /*left: 0;*/
  cursor: text;
  font-size: 75%;
  top: -1.5em;
  z-index: 3;
  line-height: 1;
  color: #000;
  white-space: nowrap;
  padding: 0 1px; }
```

2) in /public_html/layouts/resources/Fields.js: MultiEmail.addEmail() I added 
```JS
newField.find('span.far').removeClass('fa-check-square').addClass('fa-square');
```

3) in /public_html/layouts/resources/Fields.js: MultiEmail.addEmail I modified MultiEmail.toggleCheckBox() in
```JS
toggleCheckBox(element) {
  if ($(element).is(":checked")) {
    $(element).attr('checked', 'checked');
    element.closest('label.js-label-checkbox')
      .eq(0).find('span.far').eq(0)
      .removeClass('fa-square').addClass('fa-check-square');
    //element.closest('label.js-label-checkbox')
    //  .eq(0).find('svg.svg-inline--fa').eq(0)
    //  .removeClass('fa-square').addClass('fa-check-square');
  } else {
    $(element).removeAttr('checked');
    element.closest('label.js-label-checkbox')
      .eq(0).find('span.far').eq(0)
      .removeClass('fa-check-square').addClass('fa-square');
    //element.closest('label.js-label-checkbox')
    //  .eq(0).find('svg.svg-inline--fa').eq(0)
    //  .removeClass('fa-check-square').addClass('fa-square');
  }
}
```
4) in /public_html/layouts/basic/styles/Main.css:2255, ".form-inline .form-group" with "flex: 0 1 auto;"

**=== DetailView**

1. => 2.
2. it displays like editview. email without consent are not clickable. just a close litle icon: red-lock / green tick
getDisplayValue function in /modules/Vtiger/uitypes/MultiEmail.php (in listview, column width is preserved: no wider than max chars length)
```PHP
...
foreach ($value as $item) {
  if($item['o']) {
    $emails[] = parent::getDisplayValue($item['e'], $record, $recordModel, $rawText, $length) . '<span class="fas fa-check" style="color: green; margin-left: 8px;" title="' . \App\Language::translate('LBL_CONSENT_TO_SEND') . '"></span>';
  } else {
    $emails[] = \App\Purifier::encodeHtml(App\TextParser::textTruncate($item['e'])) . '<span class="fas fa-lock" style="color: red; margin-left: 8px;"></span>';
  }
}
return implode('<br>', $emails);
...
```

in getListViewDisplayValue (emails stacked)
```PHP
return strip_tags($this->getDisplayValue($value, $record, $recordModel, $rawText), '<br>');
```

3) disabled ajax inline editing. in /modules/Vtiger/uitypes/MultiEmail.php added
```PHP
public function isAjaxEditable()
{
  return false;
}
```

### === XML EXPORT

1) in /modules/Vtiger/models/ExportToXml.php isCData function, added "multiEmail" as CDATA field
```PHP
public function isCData($name, $customColumns = [])
{
  if ($customColumns) {
    return array_key_exists($name, $customColumns);
  }
  if (($fieldModel = $this->moduleFieldInstances[$name] ?? false) && in_array($fieldModel->getFieldDataType(), ['text', 'multiEmail'])) {
    return true;
  }
  return false;
}
```

=============================================

below restyled field layout
multiemaildetailnew
multiemaileditnew

### CODE REVIEW

**Comment by mariuszkrzaczkowski** [comment]
> Before explain issues, I have a question: what's better?
> 
> 1) I create directly a pull request with issue explanation
> 2) Before pull request creation, I have to create an issue and then I create a pull request
> 3) I create an issue and wait your agree about pull request creation

1

File: modules/Vtiger/uitypes/MultiEmail.php [hunk 1]
```diff
@@ -126,4 +130,10 @@ public function getOperators()
 	{
 		return ['c', 'k', 'y', 'ny'];
 	}
+
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/uitypes/MultiEmail.php]:
```suggestion
		/**
	 * {@inheritdoc}
	 */
```

File: public_html/layouts/basic/styles/Main.css [hunk 2]
```diff
@@ -3211,8 +3211,8 @@ textarea.form-control {
     display: -ms-flexbox;
     display: flex;
     -webkit-box-flex: 0;
-        -ms-flex: 0 0 auto;
-            flex: 0 0 auto;
+        -ms-flex: 0 1 auto;
```
Comment (poradz) [inline public_html/layouts/basic/styles/Main.css]: We don't edit Main.css file directly and we avoid to edit bootstrap components. You can overwrite though bootstrap classes by adding class in tpl file which overwrite bootstrap styles (in this case you can add utility class in _Utilities.scss with prefix u- with css flex behaviour you want, to overwrite form-group class, e.g. add class 'u-flex-default' to div with 'tpl-Base-Edit-Field-MultiEmailValue').

File: public_html/layouts/resources/Fields.js [hunk 3]
```diff
@@ -1478,18 +1479,20 @@ window.App.Fields = {
 		 */
 		toggleCheckBox(element) {
 			if ($(element).is(':checked')) {
```
Comment (poradz) [inline public_html/layouts/resources/Fields.js]: element is already jQuery object so it doesn't need to be wrapped by $()

File: public_html/layouts/resources/Fields.js [hunk 4]
```diff
@@ -1478,18 +1479,20 @@ window.App.Fields = {
...
+				$(element).attr('checked', 'checked');
 				element
 					.closest('label.js-label-checkbox')
 					.eq(0)
-					.find('svg.svg-inline--fa')
+					.find('span.far')
```
Comment (poradz) [inline public_html/layouts/resources/Fields.js]: We avoid references to html tags, instead add class with js- prefix for reference in js

**Comment by mariuszkrzaczkowski** [comment]
https://github.com/YetiForceCompany/YetiForceCRM/pull/10874

File: public_html/layouts/basic/styles/Main.css (hunk 2 above)
Comment (davide-alghi) [reply public_html/layouts/basic/styles/Main.css]: yes Tomasz,
it's better.

File: public_html/layouts/resources/Fields.js (hunk 3 above)
Comment (davide-alghi) [reply public_html/layouts/resources/Fields.js]: I don't remember why. Maybe some strange behavior in some browser when I cross-system / cross-browser tested... it's been a long time.
yes it's a jQuery obj

File: public_html/layouts/resources/Fields.js (hunk 4 above)
Comment (davide-alghi) [reply public_html/layouts/resources/Fields.js]: I will try to keep it in mind :D

**Pull request merged**